Multiple keyboard layout support

RandyTheDev opened the pull request:
Added the ability to automatically detect a user's keyboard layout. If a player uses a non-QWERTY keyboard layout, the plugin will intelligently adjust which keys it lights depending on the layout.

## Full Changelog
### New Features
- Supports any keyboard layout.
- The Quickload button will now only be illuminated if you have quick-loading enabled.

### Other enhancements
- The KSP Logo lighting scheme has been tuned for Logitech Keyboard layouts.
- The keyboard lighting effects will no longer disappear while waiting on loading screens.

### Housekeeping
- Extensive refactor
- Compiled against the KSP 1.4 API.
- Uses a newer version of the Logitech SDK.
- There are some basic unit tests.

## Next Steps
- Bringing this version back up to feature parity with KSP-Chroma, then going beyond.

RandyTheDev committed:
Bumped KSP Version in Netkan Config to 1.4.0

Works anyway, might as well.

RandyTheDev committed:
Modified .gitignore for Jetbrains Rider

RandyTheDev committed:
Updated Logitech SDK

RandyTheDev committed:
Added NUnit as a package.

It will be used for unit tests.

RandyTheDev committed:
Moved the Logitech code to its own directory

Also made the Logitech API wayyy more C#-y.

RandyTheDev committed:
More Refactoring

Really aggressively moving things around, hopefully it should result in
an easier to maintain project with greater scope for extensibility and
other hardware vendors.

RandyTheDev committed:
Refactoring for multi OS support.

Not for today, but it would be nice in the future. Let's not rule
anything out yet :)

RandyTheDev committed:
Windows Layout Provider Implementation

Uses the Win32 API's Keyboard Layout methods to calculate the mappings
from a user's native layout to US-QWERTY.

RandyTheDev committed:
Added initial unit tests

They check that the numpad is still the numpad. Also I renamed the
Logitech keys enum to `ScanCode` because it seemed like a good idea.

RandyTheDev committed:
Added Temporary Layouts for testing.

Temporary layouts will load a keyboard layout for testing purposes.
Once the test is complete, if the layout is no longer needed, the
layout will be unloaded. The user's layout will be returned once the
temporary layout is disposed of.

A Windows implementation has been added with this initial prototype.
This feature has been implemented in a way that keeps the door open for
cross-platform compatibility in the future.

I also deleted a bunch of stuff in the `User32` imports class, as it
was not being used.

RandyTheDev committed:
Added more layout tests

Included are tests that check the mappings of both the QWERTY and
DVORAK layout, using the `ITemporaryLayout` interfaces defined in the
previous commit!

RandyTheDev committed:
Updated main plugin class

Now uses all of my stuff.

RandyTheDev committed:
Updated ColorScheme to support any keyboard layout.

ColorScheme may now have color schemes defined for both mapped user keys
or keys as they would appear on a QWERTY board.

RandyTheDev committed:
Housekeeping

* Updated copyright dates.
* Enabled some commented out KeyCode -> ScanCode conversions.
* Modified plugin lifecycle so it is always on.
* Added docs and increased re-use.

RandyTheDev committed:
Modified the Logo to suit Logitech keyboards better

Razer Blackwidows have a funky spacebar size that threw off the design.
It's now re-calibrated for the Logitech G910!

RandyTheDev committed:
Only light up the Quickload button if enabled.

No longer light up the quickload button if you have it disabled in your
difficulty settings. Hopefully now it will be less tempting!

                                                      ... hopefully ...

RandyTheDev commented:
Hey don't merge this yet, Logitech Developer support got in contact with me and I think I have a better way of doing multi-layout support 😁

RandyTheDev committed:
Implemented runtime loading.

I've implemented runtime loading. The Logitech DLLs no longer need to
be in the KSP root folder, instead, they only need to be in the
relative folder Logitech/x64 or Logitech/x86. The code automatically
determines which version to use. Also renamed the `ScanCode` enum to
`KeyName` after talking with the Logitech SDK team.

RandyTheDev commented:
So Logitech's approach was a dead end... On the plus side, I implemented dynamic loading of the library, so KSP can now find the x64 or x86 version of the library depending on the runtime. The libraries should be installed as follows:

- KSP
  - GameData
    - KSPLogiRGB
      - Logitech
        - x86
          - LogitechLedEnginesWrapper.dll
        - x64
          - LogitechLedEnginesWrapper.dll

RandyTheDev commented:
I also still have an idea on how to do multi layouts properly.

battlemoose commented:
@RandyTheDev This looks great! You've obviously put in a lot of effort. Are you ready for me to merge at this point?

RandyTheDev committed:
Saving a branch I cannot follow.

Was trying to manually read keyboard layouts, which, while doable.
Is stupid. I'd be much better off using the existing APIs.

RandyTheDev committed:
Reverted incorrect path and added AZERTY tests

Ironically Unity and KSP don't handle multi layouts very well either,
so doing it properly would have broken everything. Actual key layouts
have been tested in KSP, seem to work okay.

RandyTheDev committed:
Added a hack to fix KSP logo on UK keyboards

Not tooo hacky, but hacky nonetheless. Everything would be fine if
Unity had better keyboard support *grumbles*

RandyTheDev committed:
Updated CKAN Installation Script

RandyTheDev committed:
Licensing my individual contributions under APL-2.0

I've licensed my individual non-KSP-specific contributions under the
Apache License. This is for a few specific reasons:

1. I would like my code to be usable by indie developers looking to
   implement RGB lighting in their own games.
2. I am considering spinning off the keyboard layout handling code into
   a separate library licensed under APL-2.0.
3. The mod itself will still be licensed under GPLv3, any KSP-specific
   code should and will remain licensed under the GPLv3. This will
   protect against Take-Two using the code without permission, or those
   attempting to sell a pre-packaged version of the mod.

I believe this strikes a fair balance between protecting the work of
contributors and opening up the source to indie developers. The "crown
jewels" of this plugin is and always will be the specific KSP
integrations, which will remain to be protected under the stricter
terms of the GPLv3 license.

RandyTheDev commented:
Hey @battlemoose!

I believe I'm ready now. I believe the layout handling code is as good as KSP will allow it to be! Please be aware of the commit where I licensed my individual contributions under the Apache license, which is GPL compatible. Commit: ada45f7

Information on GPL compatibility: https://www.gnu.org/licenses/license-list.html#apache2

RandyTheDev committed:
Updated README.md

RandyTheDev commented:
Also, make sure you download the latest version of the Logitech APIs, so that the release zip has a structure copying the one described above.

RandyTheDev commented on KSP-LogiRGB/KSP-LogiRGB/Tests/LayoutProviderTests.cs:
```diff
@@ -183,7 +186,7 @@ public void TestDvorak()
                 AssertMapped(KeyCode.R, KeyCode.O);
                 AssertMapped(KeyCode.L, KeyCode.P);
                 AssertMapped(KeyCode.Slash, KeyCode.LeftBracket);
-                AssertMapped(KeyCode.Plus, KeyCode.RightBracket);
```
While the Dvorak keycode is a plus symbol, in KSP, it is registered as a `KeyCode.Equals`
